hub/models.py

```diff
@@ -3303,6 +3330,480 @@ def get_import_data(self):
+        return self.get_sheet(sheet_name="MAPPED_LOOKUP_SHEET")
+
+    @property
+    def webhook_sheet(self):
+        return self.get_sheet(sheet_name="MAPPED_WEBHOOK_SHEET")
+
+    @cached_property
+    def headers(self) -> list[str]:
+        result = (
+            self.spreadsheets.values()
+            .get(spreadsheetId=self.spreadsheet_id, range=f"{self.sheet_name}!1:1")
+            .execute()
+        )
+        return result["values"][0]
+
+    def get_sheet(self, sheet_name: str):
+        return next(
+            (
+                sheet
+                for sheet in self.spreadsheet["sheets"]
+                if sheet["properties"]["title"] == sheet_name
+            ),
+            None,
+        )
+
+    def remote_url(self) -> str:
+        sheet = self.sheet
+        return f"{self.spreadsheet['spreadsheetUrl']}?gid=${sheet['properties']['sheetId']}"
+
+    def healthcheck(self):
+        if self.headers:
+            return True
+        return False
+
+    def field_definitions(self):
+        header_row = self.headers
+        return [
+            self.FieldDefinition(label=column, value=column) for column in header_row
+        ]
+
+    def record_url_template(self):
+        return self.remote_url()
+
+    def record_url(self, record_id: str, record_data: dict):
+        return self.remote_url()
+
+    async def fetch_one(self, member_id):
+        rows = await self.fetch_many([member_id])
+        return rows[0] if rows else None
+
+    async def fetch_many(self, id_list: list[str]):
+        row_numbers = self.fetch_row_numbers_for_ids(id_list)
+        if not row_numbers:
+            return []
+        return self.get_rows(row_numbers)
+
+    async def fetch_all(self):
+        return self.get_data()
+
+    def get_data(self) -> list[dict]:
+        result = (
+            self.spreadsheets.values()
+            .get(spreadsheetId=self.spreadsheet_id, range=self.sheet_name)
+            .execute()
+        )
+        data = result["values"]
+        headers = data[0]
+        return [dict(zip(headers, row)) for row in data[1:]]
+
+    def get_rows(self, row_numbers: list[int]) -> list[dict]:
+        last_column = google_sheets.column_index_to_letters(len(self.headers) - 1)
+        ranges = [
+            f"{self.sheet_name}!A{row_number}:{last_column}{row_number}"
+            for row_number in row_numbers
+            if row_number
+        ]
+        result = (
+            self.spreadsheets.values()
+            .batchGet(spreadsheetId=self.spreadsheet_id, ranges=ranges)
+            .execute()
+        )
+        rows = []
+        for value_range in result["valueRanges"]:
+            rows = rows + value_range["values"]
+
+        records = [dict(zip(self.headers, row)) for row in rows]
+        logger.debug(
+            f"Google Sheet source {self} got rows {records} for row numbers {row_numbers}"
+        )
+        return records
+
+    def fetch_row_numbers_for_ids(self, id_list: list[str]):
```

Comment:
An alternative to `fetch_row_numbers_for_ids` and co. could be to just ingest the whole spreadsheet and do the lookup in pandas.

An argument in favour of this is that it keeps the spreadsheet clean and prevents user error.

Argument against is memory usage, I guess.

Wdyt?